content/subtleties.md

```diff
@@ -14,37 +14,39 @@ Also see Pierre Weis's [Frequently asked Questions about Caml](http://caml.inria
-  in *Real World OCaml*,
-  [Section 5.1 Weak polymorphism and mutation](http://caml.inria.fr/pub/docs/manual-ocaml/polymorphism.html#sec51) 
-  in the OCaml Manual, [Common Error Messages](http://ocaml.org/learn/tutorials/common_error_messages.html#The-type-of-this-expression-contains-type-variables-that-cannot-be-generalized) at ocaml.org, and ["My function is not polymorphic (enough) ?"](http://caml.inria.fr/pub/old_caml_site/FAQ/FAQ_EXPERT-eng.html#eta_expansion) and subsequent sections in "Frequently asked Questions about Caml".
+  Part of the answer is that `'_weak1`, `'_weak2`, etc., which are called "weak type variables", are roughly speaking, only *temporarily* [polymorphic](http://ocaml.org/learn/tutorials/basics.html#Polymorphic-functions) types (unlike `'a`, etc.).  If an expression type includes a weak type variable, the compiler will permanently substitute the variable with a regular concrete (i.e., monorphic) type (e.g. `float list`) the first time that it sees an application of the function.
+  
+  In OCaml and other ML-style languages that support mutability, only _values_ can have polymorphic types, hence the name -- "value restriction". The underlying reason, is that if an expression is mutable (an since we don't know its type, we must assume that it is mutable), then it could be assigned multiple times, each time with a value of a different type, thus we can't guarantee that the type of a value would be preserved. That will eventually break the type soundness. 
+  
+  Originally, values were defined as syntatic constants. I.e., something that is a static constant on the source level, is considered a value, e.g., `1`, `"hello"`, `None`, `fun x -> x`. However, in OCaml this restriction is relaxed, and besides syntactic constants OCaml allows some other classes of expressions to be generalized (i.e., to have the polymoprhic type). First of all, OCaml allows polymorphism for _non-expansive_ expressions, where an expression is non-expansive if it doesn't have any observable side-effects during the evaluation. That basically extends the notion of constant from purely syntactic definition (something that looks like a constant, is a constant) to semantic definition (something that acts like a constant, is a constant). For example, this restiction allows the following expressions to have polymorphic type: `lazy None`, ` let _ = ref None in 1`, `assert false`. But that's not all, OCaml is able to prove that it is safe to generalize certain type variables, by performing their variance analysis. It was [observed]((https://caml.inria.fr/pub/papers/garrigue-value_restriction-fiwflp04.pdf)) that there is a nice interaction between subtyping, variance, and value restriction. If a value occurs only in covariant positions, then it can be safely generalized and given the polymorphic type. With this rule, it is possible to give the polymorphic type to the following expressions: `fun () -> []`, `fun () -> ref None`. 
```

Comment:
Because this line is so long, it is hard for me to comment on individual portions of it. I would still recommend line breaking around 70 columns or so.

1. I would give an example of a non-expansive expression vs an expansive one.
2. I would break the paragraph into two paragraphs at "But that's not all".
3. It would be useful to explain what generalizing a type variable (or performing variance analysis) might mean, or link to some explanation.
4. You should link to an explanation of covariance/contravariance, or we should write an explanation of it somewhere else in the FAQ. It comes up a lot and it's exceptionally confusing to newcomers. (The only reason I understood it immediately is because of my mathematics background.)